Add a new `swift-package-collections` CLI

neonichu opened the pull request:
This implements the proposed CLI for package-collections.

Requires #3028

neonichu commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+
+private extension ParsableCommand {
+    var collections: PackageCollectionsProtocol {
+        fatalError("not implemented")
```
We have to figure out how to get the necessary configuration and how to choose the storage implementation etc.

tomerd replied:
cc @abertelrud: this is the practical aspect of the question I was asking about the DI pattern we follow in SwiftPM. If there is none, we could just create the "correct" instances here 

=> Configuration should have a default ctor that will read the default config: 
=> Storage we want to use the SQLite and File implementation , make sure to close the storage when done!

abertelrud replied:
If there is a pattern in SwiftPM, I haven't found it.  So I agree, we should establish one and document it.

> => Configuration should have a default ctor that will read the default config:
> => Storage we want to use the SQLite and File implementation , make sure to close the storage when done!

I'm not sure how to translate that to what the API would look like in this case.  What component would make the choice of storage backend in this case?

tomerd replied:
the api would look the same. the call site will need to make the decision on what concrete impls of configuration and storage to use, but we would include "default impls" which would be the natural choice. we can also add an empty ctor that uses these default impls if we think its worth it

neonichu replied:
I think an empty initializer would be nice, because then clients don't need to explicitly know about things like the SQLite implementation etc.

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+private extension JSONEncoder {
+    func print<T>(_ value: T) throws where T : Encodable {
+        if #available(macOS 10.13, *) {
+            self.outputFormatting = [.prettyPrinted, .sortedKeys]
```
I think we should add `withoutEscapingSlashes`.

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+            }
+
+            let modules = version.targets.compactMap { $0.moduleName }.joined(separator: ", ")
+            let platforms = optionalRow("Supported Platforms", version.verifiedPlatforms?.map { $0.name }.joined(separator: ", "))
```
We might want to change the wording here "supported" -> "verified" to not confuse users with similar properties in `Package.swift` (they are different).

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+            let collections = try await { self.collections.listCollections(identifiers: nil, in: profileOptions.usedProfile, callback: $0) }
+
+            if jsonOptions.json {
+                try JSONEncoder().print(collections)
```
Need `withoutEscapingSlashes` I think

neonichu replied:
We'll get it through changing the implementation of `print` itself, as you mentioned above.

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+        @OptionGroup
+        var jsonOptions: JSONOptions
+
+        mutating func run() throws {
```
Do all the `run()`s really need to be `mutating`?

neonichu replied:
Yah, that's how this method is defined in the `ParsableCommand` protocol.

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+    @Option(name: .long, help: "Profile to use for the given command")
+    var profile: String?
+
+    var usedProfile: PackageCollectionsModel.Profile {
```
`usedProfile` can be optional. The API impl accepts `nil`.

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+
+            guard let collection = collections.first(where: { $0.source == source }) else {
+                throw CollectionsError.noCollectionMatchingURL(collectionUrlString)
+            }
```
I am not sure we have to throw in this case, but then the extra `listCollections` is inexpensive so I am not against doing the extra check and letting user know of the error either.

tomerd replied:
right now the business logic reports "happy" if the collection does not exist. this is typical for delete operations

neonichu replied:
I think that makes sense from the API perspective, but in CLI tools, it's common to error in such cases, e.g.

```
$ rm does-not-exist
rm: does-not-exist: No such file or directory
```

abertelrud replied:
Agreed that silent success seems good for an API, but for CLI it might be good to at least have a warning that the key is already non-existent.

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+        static let configuration = CommandConfiguration(abstract: "Get metadata for a configured collection")
+
+        @Argument(help: "URL of the collection to describe")
+        var collectionUrl: String?
```
Should this be `String`?

yim-lee commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+        @OptionGroup
+        var profileOptions: ProfileOptions
+
+        mutating func run() throws {
```
This can use the `getCollection` API--it fetches locally first and fetches the remote collection if not found

tomerd replied:
+1 that was the original intent I believe

tomerd commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+                let results = try await { collections.findTargets(searchQuery, searchType: .exactMatch, collections: nil, profile: profileOptions.usedProfile, callback: $0) }
+
+                results.items.forEach {
+                    $0.packages.forEach {
```
this may show duplicates since we are not grouping by collection so maybe flatten into a set?

tomerd replied:
also today this will search across *all* targets not just modules. should we change the API so it allows searching only on modules?

abertelrud replied:
That seems very reasonable.

neonichu replied:
I'm not sure I follow, what would be a target that isn't a module?

abertelrud replied:
A test target is generally not an importable module.

abertelrud replied:
It might be technically importable but doesn't seem like the kind of thing that should be offered in a `import` completion.

tomerd replied:
if we wanted to avoid the confusion from the term "module" vs " target" we can have the API take a targetType filter?

neonichu replied:
We could also rephrase in terms of "library target"

abertelrud replied:
Are binary targets considered to be libraries?

neonichu replied:
I would consider any target that's part of a library product a library target, so binaries as well.

abertelrud replied:
Sounds good.  Rule of thumb should be "anything that can be `import`ed", I think.

tomerd commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+private extension JSONEncoder {
+    func print<T>(_ value: T) throws where T : Encodable {
+        if #available(macOS 10.15, *) {
+            self.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
```
I think prettyPrinted was supported prior to 10.15?

abertelrud replied:
I think so as well.  Also, we're accumulating a lot of cases where we check the macOS availability for this particular purpose (specifically for JSON formatting options).  I wonder if this is worth collecting into one place where we apply the standard formatting that we always seem to want (basically "as readable as possible").

neonichu replied:
Yes, it is `.withoutEscapingSlashes` that isn't. I don't think it's worth the complexity of multiple availability checks since SwiftPM does not run on any macOS older than 10.15 anyway, but I can break it up into multiple statements if desired.

abertelrud replied:
If SwiftPM requires macOS 10.15 then we should just make that the minimum deployment target in the package manifest and have one #if check.

neonichu replied:
We can do that, but so far we intentionally avoid it to allow clients of libSwiftPM to use 10.10. We set 10.15 in `bootstrap` (see https://github.com/apple/swift-package-manager/blob/main/Utilities/bootstrap#L28), so the actual CLI tools have a 10.15 minimum deployment target.

tomerd replied:
> but so far we intentionally avoided it to allow clients of libSwiftPM to use 10.10

are there specific clients in mind?

neonichu replied:
https://github.com/apple/swift-package-manager/pull/2710#issuecomment-618731081

neonichu replied:
This will be addressed by adopting https://github.com/apple/swift-package-manager/pull/3071

tomerd replied:
I think this ^^ still stand. we should change to use the makeWIthDefaults and get rid f the setter here?

neonichu replied:
Yep, thanks for noticing.

tomerd commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+            let identity = PackageReference.computeIdentity(packageURL: packageUrl)
+            let reference = PackageReference(identity: identity, path: packageUrl)
+
+            let result = try await { self.collections.getPackageMetadata(reference, profile: profileOptions.usedProfile, callback: $0) }
```
hmm I think we may have a mistake in the API here: profile does not seem relevant for this call

yim-lee replied:
Profile is relevant. A package may show up in more than one collection and possibly with different versions. The API can either the first package found (based on ordering of collections within a profile) or "merged" data.

We can't really do "local read first and remote fetch if needed" as we do with collection (i.e., preview), because remote doesn't have all the package metadata we need (e.g., targets, products), therefore the requested package must belong to a collection.

tomerd replied:
good point!

abertelrud commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+
+            return """
+                \(version.version)
+                Package Name: \(version.packageName)
```
Should this be quoted?

neonichu replied:
https://github.com/apple/swift-package-manager/pull/3030#discussion_r520211883

abertelrud replied:
Agreed.

tomerd commented:
> I'm not sure how to translate that to what the API would look like in this case. What component would make the choice of storage backend in this case?

the api would look the same. the call site will need to make the decision on what concrete impls of configuration and storage to use, but we would include "default impls" which would be the natural choice. we can also add an empty ctor that uses these default impls if we think its worth it

neonichu commented on Package.swift:
```diff
@@ -188,6 +188,10 @@ let package = Package(
             /** Runs an executable product */
             name: "swift-run",
             dependencies: ["Commands"]),
+        .target(
```
We get an implicit one and it looks like we are using that for all executables so far.

abertelrud replied:
Ok, makes sense.  It's a bit odd, but that will be ironed out when we eventually unify products and targets.

neonichu commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+            let packages = collection.packages.map { "\($0.repository.url)" }.joined(separator: "\n")
+
+            print("""
+                Name: \(collection.name)
```
Right now the proposal doesn't propose quoting anything, but we could if we wanted to.

abertelrud replied:
Is there a parsable output format as well?  If so, that would be where things should be quoted.  The display format doesn't need to be.

neonichu replied:
Most commands have a `--json` option to alternatively emit JSON for integration with other tools.

abertelrud commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+        var profileOptions: ProfileOptions
+
+        mutating func run() throws {
+            let collections = try await { self.collections.refreshCollections(in: profileOptions.usedProfile, callback: $0) }
```
Should these all be `tsc_await` now?

abertelrud commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+        var profileOptions: ProfileOptions
+
+        mutating func run() throws {
+            guard let collectionUrlString = collectionUrl, let collectionUrl = URL(string: collectionUrlString) else {
```
Nit: it might be nice to distinguish the "missing URL" vs "malformed URL" cases.

neonichu replied:
I have solved this by making all arguments non-optional since argument-parser already takes care of error handling.

abertelrud commented on Sources/Commands/SwiftPackageCollectionsTool.swift:
```diff
@@ -0,0 +1,358 @@
+
+        let jsonData = try self.encode(value)
+        let jsonString = String(data: jsonData, encoding: .utf8)!
+        Swift.print(jsonString)
```
Is this just for debugging?  Otherwise, is sending to stdout always the right thing?  JSON on stdout is often problematic since other output will interfere (even debug output emitted while developing SwiftPM itself).  If this is intended to be an output format it would be good to have a command line flag for where to put the JSON.

neonichu replied:
Yes, this is identical to existing SwiftPM CLI features which emit JSON. If that's a concern, I would prefer to holistically address that for all commands in a separate PR than partially just for the new commands.

abertelrud replied:
Makes sense.  Eventually we should address it for all of them.  Scraping stdout is never a good idea.

neonichu commented:
@swift-ci please smoke test

neonichu commented:
@swift-ci please smoke test

neonichu commented:
- Fixed output formatting (we should be using `PlainTextEncoder` in the future, but it isn't quite obvious to me how we would omit optional/empty content, so I fixed up my handwritten output for now)
- Updated naming of the CLI tool
- Removed the tool from install in `bootstrap` since we do not want to ship it as part of 5.3

neonichu commented:
@swift-ci please smoke test

tomerd reviewed (approved):
looks good, one small comment on JSON encoding

neonichu commented:
Some commands were also missing JSON support, fixed that as well now.
